src/Compilers/CSharp/Portable/Syntax/SyntaxTreeDiagnosticEnumerator.cs

```diff
@@ -11,136 +13,120 @@ namespace Microsoft.CodeAnalysis.CSharp
-                    }
+                // Weren't able to continue with this node. Pop it so we continue processing its parent.
+                stack.Pop();
+            }
 
-                    _stack.UpdateSlotIndexForStackTop(slotIndex);
-                    _stack.PushNodeOrToken(child);
+            yield break;
+
+            bool tryContinueWithThisNode(GreenNode node)
+            {
+                // SlotCount is 0 when we hit tokens or normal trivia. In this case, we just want to move past the
+                // normal width of the item.  Importantly, in the case of tokens, we don't want to move the
+                // full-width.  That would make us double-count the widths of the leading/trailing trivia which
+                // we're walking into as normal green nodes.
+                if (node.SlotCount == 0)
+                {
+                    position += node.Width;
                 }
                 else
                 {
-                    if (node.SlotCount == 0)
+                    var nextSlotIndex = stack.Top.SlotIndex;
+                    while (++nextSlotIndex < node.SlotCount)
                     {
-                        _position += node.Width;
-                    }
+                        var child = node.GetSlot(nextSlotIndex);
+                        if (child == null)
+                            continue;
+
+                        // If the child doesn't have diagnostics anywhere in it, we can skip it entirely.
+                        if (!child.ContainsDiagnostics)
+                        {
+                            position += child.FullWidth;
+                            continue;
+                        }
+
+                        stack.Top.SlotIndex = nextSlotIndex;
+                        stack.PushNodeOrToken(child);
 
-                    _stack.Pop();
+                        return true;
+                    }
                 }
-            }
 
-            return false;
+                // Weren't able to continue with this node. Pop it so we continue processing its parent.
+                return false;
+            }
         }
 
-        /// <summary>
-        /// The current diagnostic that the enumerator is pointing at.
-        /// </summary>
-        public Diagnostic Current
+        private struct NodeIteration(GreenNode node)
         {
-            get { Debug.Assert(_current is object); return _current; }
+            /// <summary>
+            /// The node we're on.
+            /// </summary>
+            public readonly GreenNode Node = node;
+
+            /// <summary>
+            /// The index of the child of <see cref="Node"/> that we're on. Initially at -1 as we're pointing at the
+            /// node itself, not any children.
+            /// </summary>
+            public int SlotIndex = -1;
+
+            /// <summary>
+            /// We'll hit nodes multiple times.  First, when we run into them, and next, when we've processed all their
```

Comment:
It looks like we will hit a node once for each of its "iterable" children. Is that right? Since we will come back up and resume the search for the "next iterable child".

Reply:
There isn't a need to delete this flag, but, does `ProcessedDiagnostics == true` strictly correlate with `SlotIndex > -1`?